parlai/core/teachers.py

```diff
@@ -1422,6 +1427,147 @@ def _setup_data(self, path):
+        if not shared:
+            self.episodes = []
+            self.num_exs = 0
+            self.label_turns = opt.get('label_turns')
+            if opt.get('conversationteacher_datafile') is not None:
+                self._setup_data(opt.get('conversationteacher_datafile'))
+        else:
+            self.episodes = shared['episodes']
+            self.num_exs = sum(len(e) for e in self.episodes)
+
+        self.id = opt['task']
+
+        self.reset()
+
+    def share(self):
+        """
+        Share the episodes.
+        """
+        shared = super().share()
+        shared['episodes'] = self.episodes
+        return shared
+
+    def num_examples(self):
+        """
+        Return the number of examples from the data.
+        """
+        return self.num_exs
+
+    def num_episodes(self):
+        """
+        Return the number of episodes from the data.
+        """
+        return len(self.episodes)
+
+    def get(self, episode_idx, entry_idx=None):
+        """
+        Get a specific example from the dataset.
+        """
+        return Message(self.episodes[episode_idx][entry_idx])
+
+    def _setup_data(self, path):
+        logging.info("[loading data from json file into task:" + path + "]")
+        self.episodes = []
+        self.num_exs = 0
+        eps = []
+        conversations = Conversations(path)
+        self.num_exs = sum(len(c) for c in conversations)
+        for conv in conversations:
+            if conv.context:
+                warn_once(
+                    'At least one of these conversations contains a context, which is not being used'
+                )
+            turns = [t for t in conv.turns if t.get('id') != 'context']
+            if len(turns) != len(conv.turns):
+                warn_once(
+                    'At least one of these conversations contains a context within the dialogue, which is being discarded'
+                )
+            turns.insert(0, {'text': '__SILENCE__'})
+            # train on odd turns as labels (turns w/ first speaker)
+            if self.label_turns in ['firstspeaker', 'both']:
+                eps = []
+                xturns = turns[::2]
+                yturns = turns[1::2]
+                for xturn, yturn in zip(xturns, yturns):
+                    turn = {}
+                    turn['text'] = xturn.get('text').strip()
+                    turn['labels'] = [yturn.get('text').strip()]
+                    turn['episode_done'] = False
+                    eps.append(turn)
+                    self.num_exs += 1
+                if eps:
+                    eps[-1]['episode_done'] = True
+                    self.episodes.append(eps)
+
+            # train on even turns as labels (turns w/ second speaker)
+            if self.label_turns in ['secondspeaker', 'both']:
+                eps = []
+                xturns = turns[1::2]
+                yturns = turns[2::2]
+                for xturn, yturn in zip(xturns, yturns):
```

Comment:
since this for loop is basically the same as repeated above, could we make a helper function here?